src/USER-PACE/ace_abstract_basis.cpp

```diff
@@ -0,0 +1,214 @@
+//        DOUBLE_TYPE sign_factor = (signbit(x) ? -1 : 1);
+//        if (wx2 > 30.0)
+//            g = 0;
+//        else
+//            g = exp(-wx2);
+//
+//        omg = 1. - g;
+//        a = abs(x);
+//        y1 = pow(omg * a, m); //+delta
+//        y2 = m * g * a;
+//        F = sign_factor * (y1 + y2);
+//
+//        DOUBLE_TYPE dg, da, dy, dy1, dy2;
+//        dg = -2.0 * w * x * g;
+//        da = sign_factor;
+//        if (abs(y1) < eps) dy = 0.;
+//        else
+//            dy = m * y1 / (omg * a); // + delta
+//
+//        dy1 = dy * (-dg * a + omg * da);
+//        dy2 = m * (dg * a + g * da);
+//        DF = sign_factor * (dy1 + dy2);
+//
+//    } else {
+//        F = m * x;
+//        DF = m;
+//    }
+//}
+
+////embedding function
+////case nemb = 1 only implementation
+////F = sign(x)*(  ( 1 - exp(-(w*x)^3) )*abs(x)^m + ((1/w)^(m-1))*exp(-(w*x)^3)*abs(x) )
+//// !! no prefactor wpre
+void Fexp(DOUBLE_TYPE x, DOUBLE_TYPE m, DOUBLE_TYPE &F, DOUBLE_TYPE &DF) {
+    DOUBLE_TYPE w = 1.e6;
+    DOUBLE_TYPE eps = 1e-10;
+
+    DOUBLE_TYPE lambda = pow(1.0 / w, m - 1.0);
+    if (abs(x) > eps) {
+        DOUBLE_TYPE g;
+        DOUBLE_TYPE a = abs(x);
+        DOUBLE_TYPE am = pow(a, m);
+        DOUBLE_TYPE w3x3 = pow(w * a, 3);
+        DOUBLE_TYPE sign_factor = (signbit(x) ? -1 : 1);
+        if (w3x3 > 30.0)
+            g = 0.0;
+        else
+            g = exp(-w3x3);
+
+        DOUBLE_TYPE omg = 1.0 - g;
+        F = sign_factor * (omg * am + lambda * g * a);
+        DOUBLE_TYPE dg = -3.0 * w * w * w * a * a * g;
+        DF = m * pow(a, m - 1.0) * omg - am * dg + lambda * dg * a + lambda * g;
+    } else {
+        F = lambda * x;
+        DF = lambda;
+    }
+}
+
+
+//Scaled-shifted embedding function
+//F = sign(x)*(  ( 1 - exp(-(w*x)^3) )*abs(x)^m + ((1/w)^(m-1))*exp(-(w*x)^3)*abs(x) )
+// !! no prefactor wpre
+void FexpShiftedScaled(DOUBLE_TYPE rho, DOUBLE_TYPE mexp, DOUBLE_TYPE &F, DOUBLE_TYPE &DF) {
+    DOUBLE_TYPE eps = 1e-10;
+    DOUBLE_TYPE a, xoff, yoff, nx, exprho;
+
+    if (abs(mexp - 1.0) < eps) {
+        F = rho;
+        DF = 1;
+    } else {
+        a = abs(rho);
+        exprho = exp(-a);
+        nx = 1. / mexp;
+        xoff = pow(nx, (nx / (1.0 - nx))) * exprho;
+        yoff = pow(nx, (1 / (1.0 - nx))) * exprho;
+        DOUBLE_TYPE sign_factor = (signbit(rho) ? -1 : 1);
+        F = sign_factor * (pow(xoff + a, mexp) - yoff);
+        DF = yoff + mexp * (-xoff + 1.0) * pow(xoff + a, mexp - 1.);
+    }
+}
+
+void ACEAbstractBasisSet::inner_cutoff(DOUBLE_TYPE rho_core, DOUBLE_TYPE rho_cut, DOUBLE_TYPE drho_cut,
+                                       DOUBLE_TYPE &fcut, DOUBLE_TYPE &dfcut) {
+
+    DOUBLE_TYPE rho_low = rho_cut - drho_cut;
+    if (rho_core >= rho_cut) {
+        fcut = 0;
+        dfcut = 0;
+    } else if (rho_core <= rho_low) {
+        fcut = 1;
+        dfcut = 0;
+    } else {
+        fcut = 0.5 * (1 + cos(M_PI * (rho_core - rho_low) / drho_cut));
```

Comment:
This core codebase (i.e. all other files except for pair_pace.*) is a part of another project that  does not involve LAMMPS::MathConst namespace. And since I checked that M_PI and  LAMMPS::MathConst ::MY_PI are both identical and both equal to 3.14159265358979323846, then I would not like to change `M_PI` -> `MY_PY` here.

Reply:
Then we have a bit of a problem. Not so much with this particular issue, but in general. This is not going to be the only request to changes to the files in the USER-PACE folder to make them conform with the LAMMPS coding style and conventions, and some of those requests *will* conflict with what you require to use the files in a different code. I will have to discuss with @athomps and @sjplimp what to do. We will need to find a compromise.

Reply:
We can accept your suggestions, but in that case the continous integration from our repos to github/LAMMPS wold be broken and that would  require manual tuning the code every time for PR.

Reply:
Please wait until I have discussed this with my colleagues.

Reply:
@yury-lysogorskiy we discussed your contribution earlier today and here is outcome:
- for all files under src/ we expect that certain minimal requirements for style and code layout are followed and also that the LAMMPS developers can make changes to reflect other global changes that we do for consistency. E.g. the MY_PI vs M_PI is because M_PI may be defined differently on different platforms (and we want consistent behavior, even if for many platforms there is no difference). There are also certain code constructs that we are trying to avoid or remove (e.g. we want LAMMPS to eventually be fully re-entrant and thread-safe, that is for the code under src/). So leaving the files in place would require to make modifications as requested, and there will likely be more, since I only pointed out a few issues that stood out.
- if you want to keep files (mostly) unmodified, you need to move them from src/USER-PACE to a new folder lib/pace and then leave only the pair_pace.h and pair_pace.cpp files in src/USER-PACE. You can look at the POEMS and the USER-COLVARS packages as examples for how to do this (both implement a fix that connects to a library in lib, the colvars one is maintained as a separate project and also bundled with NAMD, for example). The requirements for code in the pair style file as given above are still applied strictly (the pair_pace.h file is where we worry the most), but it looks to me as if it would require only minor modifications to meet them. 

Independent from that, I think the suggestions I made in addition are worth taking under consideration for your other code, and then the files can still be shared in unmodified form, since those are based on many years of maintaining a code with as many contributors and with as diverse contributions as we have for LAMMPS. We have been through several very complicated bugs that ultimately were due to unexpected side effects from header files that did things that we now discourage people to do. This is why we have started writing files to document the conventions and approaches we follow.

Reply:
@akohlmey , thank you for the explanation! `pair_pace.*` files are completely flexible for us and we could gladly accept all modifications in there. On the other hand, I would like to ask what are drawbacks of having USER-PACE in lib/pace? Less support from core LAMMPS developers when the LAMMPS core is changing?  Would that be possible to use third-party libraries under non-GPL2 license if the code is in LAMMPS/lib?

Reply:
And having USER-PACE in lib/pace, does it mean that user should do extra steps of compiling the library first and the linking it during main LAMMPS compilation?

Reply:
> @akohlmey , thank you for the explanation! `pair_pace.*` files are completely flexible for us and we could gladly accept all modifications in there. On the other hand, I would like to ask what are drawbacks of having USER-PACE in lib/pace? Less support from core LAMMPS developers when the LAMMPS core is changing?

Code in LAMMPS/src is regularly subject to static code analysis and portability testing. If we make changes to low-level interfaces we adapt code accordingly. Also we have been refactoring internal tools and are regularly adding tools to simplify repetitive tasks, so we may adapt code to take advantage of those, too. For code in LAMMPS/lib there is much less scrutiny and less effort invested. Generally, the code in lib should be fully independent from LAMMPS, i.e. not require any LAMMPS headers and only imported into the .cpp files of files in the package folder or only in header files that are only used by the .cpp files, but not included directly or indirectly outside of the scope of the package.

> Would that be possible to use third-party libraries under non-GPL2 license if the code is in LAMMPS/lib?

All code included in the LAMMPS distribution would have to comply with the LAMMPS licensing terms: GPLv2 for the public distribution and LGPLv2.1 on demand for people writing (commercial) tools on top of LAMMPS that want to link LAMMPS as a library to their tool (no code merging or modification of the LAMMPS code).  If you want to distribute code under a license not compatible with that (e.g. GPLv3 is more restrictive and thus not compatible), it can be accommodated by compiling this code in lib/package as well, but then the source code is not included and instead downloaded during compilation/configuration of the library. There is precedent for that, too, e.g. see the LATTE, KIM, or USER-PLUMED package, where the download is automatic. There are also packages where this is not automatic (e.g. QUIP, where the GAP support requires a manual download with agreeing to non-GPL compatible license terms). The the compilation/configuration of the library is manual, but then the LAMMPS package code (in src/) can be installed like any other package and imports the necessary settings from files in lib/package or through CMake variables/scripts.

Reply:
> And having USER-PACE in lib/pace, does it mean that user should do extra steps of compiling the library first and the linking it during main LAMMPS compilation?

the separate configuration/compilation of the library is true when using the build procedure with GNU make. When using CMake this step is integrated.

Essentially, this is a scale: for code in src/ the requirements are the most, but also the convenience is the largest, and the chances of the package being included when people compile LAMMPS for others (e.g. HPC centers, Linux distributions, Homebrew, etc.). With the code in lib/ this more often excludes the package (e.g. some people build LAMMPS through conventional make with: make yes-all; make no-lib). With a different license and non-automatic download chances are the smallest, and people have to make a conscious decision to download and compile everything.

As much as you have an interest to keep code in your style and manage it your way, we have a desire to keep things consistent and more maintainable for us. As LAMMPS is growing over the years, this is becoming increasingly challenging and thus we have started to be more demanding with new contributions and worked on existing code to become more consistent. From the information you provide here, my impression is that your situation is probably closest to what applies to the LATTE package, but my sense is that distributing a copy of the supporting files in lib is doable with minimal changes and that would improve the situation somewhat.

So we will have to find a compromise that can work well enough for both sides. From our perspective, having code as libraries in lib/ is an additional burden and nuisance, but having code in src/ that is not consistent with our requirements for organization, code layout, style, and conventions is the bigger problem.  Please note that there have been **very** few packages where there were consistent and regular updates to the code in the LAMMPS distribution. the KOKKOS library and the colvars library are pretty much it. In cases where we did receive infrequent updates those were usually years apart and then updating the code to match the LAMMPS conventions at the time was a significant effort, but rather small effort when considering how long there was not need to invest any effort at all.

I suggest you dig through the sections on building LAMMPS and packages in LAMMPS a bit to see what the differences are between the various examples of packages I have mentioned and what it says about compiling/configuring LAMMPS with those packages included.

Reply:
And here is another thought: Most of the utility code or portability/usability enhancements are collected in a very small number of files and headers. Since LAMMPS **is** GPL, you could pick the few things needed, include them in your project and thus make the support files shared between your distribution and LAMMPS compatible even when using those LAMMPS features/conventions in them. E.g. we include utils.h indirectly through a header that is included by the vast majority of files in LAMMPS, so those function calls are always available

Reply:
After some discussions, we decided, that having a "weak" coupling to the LAMMPS repo would be better. So, we would be happy, if the LAMMPS could pull the PACE implementation during building phase (using CMake?) from 3rd party repository (is that possible?).

We have following questions regarding this:
1. Where will pair_pace.h/cpp go? To LAMMPS/lib/USER-PACE or to LAMMPS/src/USER-PACE  ?
2. How should we organize the external repo? Which files should be there?
3. How should LAMMPS/cmake/CMakeLists.txt should be modified? How will do these modifications?

Reply:
pair_pace.cpp and pair_pace.h will have to go into src/USER-PACE and thus also need to conform to our requirements.
also src/USER-PACE would need to have an Install.sh file that integrates the settings from the lib folder into the conventional build process. there would have to be a lib/pace folder for building/interfacing to the library.

how you organize your repo is up to you. a better way to describe what you would have to do is to point you to examples.
which of the examples is relevant depends on what build system you are using for your "PACE" external package.

1. If you are using CMake to build it, you can look at the KIM package
2. If you are using autoconf/automake, you can look at the USER-PLUMED package
3. If you are using neither (just a plain Makefile), you can look at the VORONOI or QUIP package

The following additional modifications are needed:
- lib/pace: needs one or more Makefile.lammps, a README and an Install.py file
- src/Makefile: needs support for the USER-PACE package and compiling via lib-pace. Also the package needs to be added to the PACKEXT and PACKLIB variables
- cmake/Modules/Packages/USER-PACE.cmake needs to be added
- cmake/CMakeLists.txt needs to include it in case the package is enabled.
- doc/src/Build_extra.rst needs to include the build instructions for CMake and traditional make
- doc/src/Package_user.rst and doc/src/Package_details.rst needs to contain relevant information and links 

Because of the different build systems, there are different steps required, but each of the examples I've pointed out are tested and used regularly and thus should be working sufficiently well. Mind you the QUIP package is set up in such a way, that no automatic download is possible and a manual download and compilation is required. So that is the least preferred and least convenient option.

We are happy to provide more details, but that requires that you first have something that already is following either of the suggested variants, so that we don't have to discuss in all generality.

I would also suggest to close this pull request here, leave all files in place, but then start a new branch from the current master and then move the few things over you need to retain, add the build environment files and then start working on getting it to do what it should.

Reply:
@akohlmey Thank you for the detailed explanation. Sorry to coming back to this after long delay.
 I have another question then: does it really necessary to have USER-PACE as a library compiled in lib/USER-PACE? or could we download the sources and put them into src/USER-PACE with  src/USER-PACE/Install.sh (when make yes-user-pace is called?) ?

Reply:
> @akohlmey Thank you for the detailed explanation. Sorry to coming back to this after long delay.
> I have another question then: does it really necessary to have USER-PACE as a library compiled in lib/USER-PACE? or could we download the sources and put them into src/USER-PACE with src/USER-PACE/Install.sh (when make yes-user-pace is called?) ?

Yes, this is necessary. What you suggest will result in inconsistent behavior and cannot be cleanly implemented into the CMake build procedure which depends on the src tree being unmodified.